xds/src/main/java/io/grpc/xds/XdsClient.java

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * An {@link XdsClient} instance encapsulates all of the logic for communicating with the xDS
+ * server. It may create multiple RPC streams for a series of xDS protocols (e.g., LDS, RDS,
+ * VHDS, CDS and EDS) over a single channel. Watch-based interfaces are provided for each set
+ * of data needed by gRPC.
+ *
+ * <p>This class should only be instantiated by the xDS resolver but can be passed to load
+ * balancing policies.
+ */
+abstract class XdsClient {
+
+  /**
+   * Data class containing the results of performing a series of resource discovery RPCs via
+   * LDS/RDS/VHDS protocols. The results may include configurations for path/host rewriting,
+   * traffic mirroring, retry or hedging, default timeouts and load balancing policy that will
+   * be used to generate a service config.
+   */
+  // TODO(chengyuanzhang): content TBD, most information comes from VirtualHost proto.
+  static final class ConfigUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class CdsUpdate {
+
+  }
+
+  // TODO(zdapeng): content TBD.
+  static final class EdsUpdate {
+
+  }
+
+  /**
+   * Config watcher interface. To be implemented by the xDS resolver.
+   */
+  @NotThreadSafe
+  interface ConfigWatcher {
+
+    /**
+     * Called when receiving an update on virtual host configurations.
+     */
+    void onConfigChanged(ConfigUpdate update);
+
+    void onError(Throwable t);
+  }
+
+  /**
+   * Cluster watcher interface.
+   */
+  interface ClusterWatcher {
+
+    void onClusterChanged(CdsUpdate update);
+
+    void onError(Throwable t);
+  }
+
+  /**
+   * Endpoint watcher interface.
+   */
+  interface EndpointWatcher {
+
+    void onEndpointChanged(CdsUpdate update);
+
+    void onError(Throwable t);
+  }
+
+  /**
+   * Starts virtual host configuration config watch for the given target name.
+   */
+  abstract void watchConfigData(String targetName, ConfigWatcher watcher);
+
+  /**
+   * Cancels virtual host configuration config watch for the given target name.
+   */
+  abstract void cancelConfigDataWatch(String targetName);
```

Comment:
This doesn't look right. For canceling the caller needs to pass the original `watcher` reference since that's the one needs canceling. If you get 2 callers watching for the same `targetName` how will you cancel them separately? Same comment for other `cancel*` functions below

Reply:
There will be only one watcher for each set of data (e.g., service config/CDS/EDS data): 
- the xDS resolver will watch on config update
- CDS LB policy will watch on CDS update
- EDS LB policy will watch on EDS update

What's the case for having multiple watchers watching on, say CDS, data?

Reply:
I cannot come up with a use case right now but may be in future when the LB code becomes quite complex there might be multiple watchers for the same name. If you are going to assume just one watcher per name (which I think we should not assume), are you going to flag errors when 2 watchers call say `watchClusterData` for the same `clusterName` and with different `watcher` instances? And with the same `watcher` instances? 

What will be ideal is the `watchClusterData` returns a `WatcherReference` (some kind of opaque ID) which is then passed to the `cancel*` function to cancel. This is cleanest even if you want to support a single watcher per name.

Reply:
Sure. I saw c++'s implementation actually supports multiple watchers for CDS/EDS (see https://github.com/grpc/grpc/blob/8619219081148b788066c00ad8328c2806d1da36/src/core/ext/filters/client_channel/xds/xds_client.h#L174-L177). We can do that as well. But I will wait for @dapengzhang0's opinion (or can make changes directly to this PR, maintainer edits enabled) as he is the implementor for those pieces. 

For service config update, c++ implementation does not support multiple watchers (see https://github.com/grpc/grpc/blob/8619219081148b788066c00ad8328c2806d1da36/src/core/ext/filters/client_channel/xds/xds_client.cc#L1192-L1198). Only the xDS resolver will be watching on service config change, currently I do no see other reasons to pass LDS/RDS/VHDS data to other components.

Reply:
> Sure. I saw c++'s implementation actually supports multiple watchers for CDS/EDS (see https://github.com/grpc/grpc/blob/8619219081148b788066c00ad8328c2806d1da36/src/core/ext/filters/client_channel/xds/xds_client.h#L174-L177). We can do that as well. But I will wait for @dapengzhang0's opinion (or can make changes directly to this PR, maintainer edits enabled) as he is the implementor for those pieces.

With multiple watchers supported, when a `cancel*DataWatch` for a name is called are you going to cancel all the watchers registered for that name? That doesn't sound right.

> 
> For service config update, c++ implementation does not support multiple watchers (see https://github.com/grpc/grpc/blob/8619219081148b788066c00ad8328c2806d1da36/src/core/ext/filters/client_channel/xds/xds_client.cc#L1192-L1198). Only the xDS resolver will be watching on service config change, currently I do no see other reasons to pass LDS/RDS/VHDS data to other components. Although weaken the specification to allow multiple watchers on config update doesn't break the behavior, we should keep the `XdsClient` API meaningful.

I thought you will have a common implementation for all the watchers. Of course you can come up with a common implementation that supports both use cases based on a flag.

Reply:
I propose 2 alternatives:

First:
```
interface WatchHandle {
}

abstract WatchHandle watchClusterData(String clusterName, ClusterWatcher watcher);

abstract void cancelClusterDataWatch(WatchHandle handle);
```
Second:
```
abstract void watchClusterData(String clusterName, ClusterWatcher watcher);

abstract void cancelClusterDataWatch(ClusterWatcher watcher);
```

Reply:
In C-core, we support multiple watchers for cluster and endpoint data, since there may be cases where we are doing a graceful swap between two LB policy trees where there are temporarily two watchers for the same cluster or endpoint data.

For example, let's say that the LDS/RDS data initially points at a single cluster X, so we configure the CDS LB policy for cluster X as the top-level policy, and that policy starts a watch for cluster X.  Now we get an LDS/RDS update that changes to using WeightedClusters to distribute the load across clusters X and Y, so we need to configure the WeightedCluster LB policy with a CDS child policy for clusters X and Y.  When we gracefully switch from the original config to the updated one, we will temporarily have two different CDS LB policies doing a watch for cluster X -- in the original config, it's the top-level policy, but in the updated config, it's a child policy.

The reason that we do not support multiple watchers for service config data is that the resolver is never swapped out, so there is no use-case where we need to do graceful switching there.  So this requirement only applies to the cluster and endpoint watchers.

Reply:
Thanks, this explanation is super clear.

Reply:
Also, is it possible that a single cluster X is ultimately used in more than one service? For example when `"xds:///x.foo.com"` and `"xds:///y.bar.com"` are separately resolved by 2 different gRPC clients they both end using cluster X so there exist 2 separate watchers for cluster X in 2 separate hierarchies? But I guess there will be 2 separate `XdsClient`s (an XdsClient per resolver) so not an issue? But it will be good to not have to make that assumption.

Reply:
There will be one `XdsClient` per channel, and the `XdsClient` instances will not share data between them, so I don't think this case is an issue.  If we ever change that part of the design, we can change this API at that point, so I don't see any reason that we need to support it now.

Reply:
I don't have strong opinion right now. My opinion is we don't even need to add the cancel methods right now, and until when we need to call cancel in any of our implementation parts then we can add the cancel methods.